### ISSUE TITLE
summary: reduce in-memory footprint

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#19073]
Summary entries currently look like this:
```c++
struct summary_entry {
    int64_t raw_token;
    bytes_view key;
    uint64_t position;
};
```

Keys are stored externally, in a separate container. We could drop the key from this structure because the token alone should be enough to do lookups, making it fixed-size and in certain tables, this would result in a significant reduction in size. The only drawback is having to adapt our lookup algorithm to deal with a potential (but very unlikely) token collision.

We also look into dropping some of the fields in `summary_ka`, which are only needed when parsing the summary from disk, like `headers` and possibly more.

**Comment by denesb** [comment]
Implementing trie-based index makes this obsolete, closing.

**Comment by mykaul** [comment]
> We also look into dropping some of the fields in `summary_ka`, which are only needed when parsing the summary from disk, like `headers` and possibly more.

Is that also irrelevant / benign?

**Comment by denesb** [comment]
> > We also look into dropping some of the fields in `summary_ka`, which are only needed when parsing the summary from disk, like `headers` and possibly more.
> 
> Is that also irrelevant / benign?

Yes, we won't have a summary at all in the new format.

**Comment by avikivity** [comment]
> Implementing trie-based index makes this obsolete, closing.

Are we implementing the trie based index?

**Comment by mykaul** [comment]
> > Implementing trie-based index makes this obsolete, closing.
> 
> Are we implementing the trie based index?

We'll try to.

**Comment by raphaelsc** [comment]
isn't the cost-benefit ratio of implementing this good enough for the short term? this probably can be implemented under a week (and it may be more backportable), while trie-based index can easily take more than 1 month (with testing).

**Comment by raphaelsc** [comment]
the trie-based index is not backportable at all, right? I think it needs a cluster feature flag. so this may be worth implementing to bring alleviation to users, until they all upgrade to a version containing trie based index (how long can that take?)

**Comment by denesb** [comment]
I don't think we need a "quick fix", there is no urgent pain in the field related to this. RAM:DISK improvement in general is not backport material, it is more of a strategic improvement.

**Comment by denesb** [comment]
Reopening, this improvement is quite cheap and it was decided we want improvements for the m format too.

**Comment by nyh** [comment]
> > Implementing trie-based index makes this obsolete, closing.
> 
> Are we implementing the trie based index?

https://github.com/scylladb/scylladb/issues/19191

**Comment by denesb** [comment]
Replaced by SCYLLADB-119.
